### ISSUE TITLE
Performance regression after updating from May 2024 version

### ISSUE THREAD

**Issue by lnxsrt** [neosr-project/neosr#102]
I had previously been using neosr from the May 2024 ([https://github.com/neosr-project/neosr/commit/c8720232448eb059567ae820e3d461d66a4aef1c](https://github.com/neosr-project/neosr/commit/c8720232448eb059567ae820e3d461d66a4aef1c)) timeframe. I generated some models that were really good for my use case and I haven't used the software since. 

Recently, I wanted to check out if there were any new cutting edge models available, so I grabbed the latest neosr. First wanted to confirm that I could get similar results from a previous ESRGAN training run I had made. However, using the exact same settings, training data, and validation data as before, the latest neosr is struggling. Below is a comparison of old neosr in orange vs new neosr in blue, and new neosr with warmup in red (noted by commit hash).

![Image](https://github.com/user-attachments/assets/5a02c6eb-e840-4817-bcda-2e70d51ed3ac)

Previously I used an initial 8e-4 learning rate with NAdam, on a multisteplr schedule. Now the model can't even get going without using a long warmup cycle, and then it will begin to diverge on the validation data within 30k iterations. I've tried everything to get a model that doesn't quickly diverge on the validation data. I've tried changing settings, optimizers (schedule-free and normal), tuning parameters (lr, decay), other models (drct, hma), etc. for weeks. I can't even get close to the performance/stability I was getting on the older version of neosr.

Maybe I'm missing something, but it feels like there is a regression somewhere. Either in neosr or the latest versions of the dependencies. It isn't an nvidia driver or hardware issue because I can rerun the old neosr (I kept the virtual env) and replicate the old results. The issue appears to be in one of the core modules/dependencies because the performance is universally bad across optimizers, model types, and settings.

Attached configs, logs, and python environments.

[train_esrnet_neosr_3638c16.toml.txt](https://github.com/user-attachments/files/19274833/train_esrnet_neosr_3638c16.toml.txt)
[train_esrnet_neosr_3638c16_warmup.toml.txt](https://github.com/user-attachments/files/19274835/train_esrnet_neosr_3638c16_warmup.toml.txt)
[train_esrnet_neosr_c872023.yml.txt](https://github.com/user-attachments/files/19274834/train_esrnet_neosr_c872023.yml.txt)

[train_train_esrnet_neosr_3638c16_20250316_112522.log](https://github.com/user-attachments/files/19274691/train_train_esrnet_neosr_3638c16_20250316_112522.log)
[train_train_esrnet_neosr_3638c16_warmup_20250316_121932.log](https://github.com/user-attachments/files/19274692/train_train_esrnet_neosr_3638c16_warmup_20250316_121932.log)
[train_train_esrnet_neosr_c872023_20240420_215027.log](https://github.com/user-attachments/files/19274693/train_train_esrnet_neosr_c872023_20240420_215027.log)

[env_3638c16.yml.txt](https://github.com/user-attachments/files/19274696/env_3638c16.yml.txt)
[env_c872023.yml.txt](https://github.com/user-attachments/files/19274699/env_c872023.yml.txt)

**Comment by neosr-project** [comment]
Hi. Unfortunately I can't replicate your issue. We have been training with the latest neosr for quite sometime now, no issues on our side. One reason I can think of is the `augmentation` config. Disable that if you're not doing real-world SR (you seem to be training with only L1 loss, so I assume you're doing bicubic only).

**Comment by lnxsrt** [comment]
Disabled augmentations and rotation. No difference. The original run using the old neosr had augmentations (I attempted to match configs exactly, some configuration options and yml->toml had changed between my git pulls).

I am doing real-world SR (I did my own degradations/subimages that better match the degradations in the inference data and use paired datasets). This is training the base model before a 2nd fine-tuning run with a GAN.

The only thing I can think to try is to start pulling commits between May 2024 and now, and figure out which one broke my particular situation.

**Comment by neosr-project** [comment]
🤔 
> The only thing I can think to try is to start pulling commits between May 2024 and now, and figure out which one broke my particular situation.

Nothing has changed that would explain that completely. Maybe `grad_clip`, as well as clipping images to maintain it in range 0-1. You can disable [`grad_clip`](https://github.com/neosr-project/neosr/wiki/Configuration-Walkthrough#grad_clip) if you wish. I'd also recommend you use other network for testing (such as SPAN). ESRGAN is only on neosr for historical reasons, it has been superseded by many networks now.
Also note that results won't match 100% regardless. Not only due to many changes that happened on neosr, but also because you're not using `manual_seed` (which enables determinism). In any case, a clean installation of neosr using the installation scripts should give you similar or better results (if using the default configurations provided on the options/ folder).
Let me know if you find the reason for your issue, I'm interested if this is something specific to your setup or if it could be something happening with other people. As I've said, many other people has been training without such issue, so it makes me think something is off with your setup.

**Comment by lnxsrt** [comment]
I have tested 6 commits between May 2024 to Sept 5, 2024 commit ([https://github.com/neosr-project/neosr/commit/dc4e3742132bae2c2aa8e8d16de3a9fcec6b1a74](url)) and so far they are working as expected. This included switching from my old environment with pytorch 2.2.2 and my current environment with pytorch 2.5.1. So, I know my new python environment is fine.

I'll continue to test up to the latest. Should find it eventually 😃

**Comment by neosr-project** [comment]
Hi, thanks for debugging this issue. This is an interesting issue, you're the only one who found it (says a lot about how ESRGAN is 'dead'). The issue seems to be with the `kaiming_normal_` init used in ESRGAN. It probably creates very large values, above 1.0. So nearly all values end up clipped by the code above.
I'll make the commit an option, so people can disable it just in case. Thanks for reporting 👍

**Comment by neosr-project** [comment]
Fixed on [`de60639`](https://github.com/neosr-project/neosr/commit/de606394efaf7065233f4acb8b7cfc727c18ab2d). Option has been documented on the [wiki](https://github.com/neosr-project/neosr/wiki/Configuration-Walkthrough#clamp) and ESRGAN defaults has been changed. Thanks again for reporting.

**Comment by lnxsrt** [comment]
Good deal. Thanks for adding the option. 

Another issue you may have noticed in my screenshot is that the dists metric is calculated differently now than in the past. I think this is also a bug. Looks like a couple of lines were swapped [here](https://github.com/neosr-project/neosr/commit/de9c33ccfb61dd0b58359b7c73a02dbd8ee9389f#diff-664773a8d050a3b03960ba558de095079be7e83de04d891f1ae45e13e2bf997fL130).

Specifically:
`dist1 = 0`
`dist2 = 0`

I think the previous position of them being outside of the loop is correct (otherwise they reset on each loop, which does not seem like the intended behavior). Making that change makes the dists metric work as before.

**Comment by neosr-project** [comment]
Another good observation, very helpful, thanks a lot. Fixed on [`aca5aa7`](https://github.com/neosr-project/neosr/commit/aca5aa75879ab9abcfb8d4d04361af3cdbcdda52).

**Comment by lnxsrt** [comment]
I ran the model out a bit further, and was still having divergence issues on the validation dataset after 15-30k iterations. I suspected it was related to another instance of clamping, and sure enough, [this](https://github.com/neosr-project/neosr/commit/aa3e5ba7b08824d4ee83fd2d865829a76fefd344) commit is the cause. After reverting these changes, I get nearly an identical match on the validation metrics as I was getting before, and the model is also subjectively improving as it should on the test set.

What is the purpose of clamping the augmented images to a range of 0 to 1?

**Comment by neosr-project** [comment]
The commit you point above is correct though...
> What is the purpose of clamping the augmented images to a range of 0 to 1?
They are already in range 0-1. However, `F.interpolate` does not guarantee it to be in range after its operation. So it's always correct to clamp it after it, making sure values are not above or bellow (which would be outside the range of an 8-bit image).

**Comment by lnxsrt** [comment]
Yep, I added back the clamp and it is working fine. The issue was the addition of `antialias=True` in the same commit.

**Comment by neosr-project** [comment]
🤔 I guess it is possible that `antialias` could cause a misalignment when the augmentation patch is merged back. I remember at the time of making this commit testing without issues, but I'll test it again asap. Thanks for investigating even further.

**Comment by lnxsrt** [comment]
I've been running tests on other models like HMA and DRCT and these fixes have resolved the divergence issues (on validation) I was having with those as well. I think the `antialias=True` on augmentations was the main issue I was having with those models.

**Comment by neosr-project** [comment]
Fixed on [`00cb24c`](https://github.com/neosr-project/neosr/commit/00cb24c2237bc917655ebc9cdb1ad0f29b063a43). I'm closing this thread for now, but feel free to open a new one at any time in case you find anything else, or send a PR to the `dev` branch. Thanks again for reporting those issues 👍
